pkg/requests/builder.go

```diff
@@ -0,0 +1,173 @@
+func (r *builder) WithMethod(method string) Builder {
+	r.method = method
+	return r
+}
+
+// WithHeader replaces the request header map with the given header map.
+func (r *builder) WithHeader(header http.Header) Builder {
+	r.header = header
+	return r
+}
+
+// SetHeader sets a single header to the given value.
+// May be used to add multiple headers.
+func (r *builder) SetHeader(key, value string) Builder {
+	if r.header == nil {
+		r.header = make(http.Header)
+	}
+	r.header.Set(key, value)
+	return r
+}
+
+// Do performs the request and returns the response in its raw form.
+// If the request has already been performed, returns the previous result.
+// This will not allow you to repeat a request.
+func (r *builder) Do() (*http.Response, error) {
+	if r.response != nil {
+		// Request has already been done
+		return r.response, nil
+	}
+
+	// Must provide a non-nil context to NewRequestWithContext
+	if r.context == nil {
+		r.context = context.Background()
+	}
+
+	req, err := http.NewRequestWithContext(r.context, r.method, r.endpoint, r.body)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %v", err)
+	}
+	req.Header = r.header
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error performing request: %v", err)
+	}
+
+	r.response = resp
+	return resp, nil
+}
+
+// Into performs the request and attempts to unmarshal the response into the
+// the given interface. The response body is assumed to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func (r *builder) Into(into interface{}) error {
+	resp, err := r.Do()
+	if err != nil {
+		return err
+	}
+
+	return Into(resp, into)
+}
+
+// JSON performs the request and attempts to unmarshal the response into a
+// simplejson.Json. The response body is assume to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func (r *builder) JSON() (*simplejson.Json, error) {
+	resp, err := r.Do()
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := getResponseBody(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := simplejson.NewJson(body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading json: %v", err)
+	}
+	return data, nil
+}
+
+// Into attempts to unmarshal the response into the the given interface.
+// The response body is assumed to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func Into(resp *http.Response, into interface{}) error {
+	body, err := getResponseBody(resp)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(body, into); err != nil {
+		return fmt.Errorf("error unmarshalling body: %v", err)
+	}
+
+	return nil
+}
+
+// getResponseBody extracts the response body, but will only return the body
+// if the response was successful.
+func getResponseBody(resp *http.Response) ([]byte, error) {
+	body, err := ioutil.ReadAll(resp.Body)
+	resp.Body.Close()
```

Comment:
Do we want to handle this error? I've had it trained into me to handle all errors (my IDE likes to yell at me about it) -- but I even seen go underlying `http` library ignoring error handling on Body.Close()

Reply:
Had a dig through the code, given we have done an `ioutil.ReadAll` just above, there's no need to check the error as the code path checks whether the body has been read, because it is at EOF, it returns a non-nil error. What we should do is check the error from the line above first, the typical pattern is to defer the response body close so I'll do that